### ISSUE TITLE
SA1003: SymbolsMustBeSpacedCorrectly flags violations in code passing the original stylecop implementation

### ISSUE THREAD

**Issue by lowkay** [DotNetAnalyzers/StyleCopAnalyzers#1300]
Howdy,

The rule for SA1003 specifies that symbols must be spaced correctly, and includes a defined list of operators for consideration. The implementation in StyleCopAnalyzers treats more cases as violations than the original StyleCop namely:

1. Colon ':' for constructor inheritance e.g.
    ```C#
    public class ClassB : ClassA {
        /// Violation here as "Operator ':' must not appear at the end of a line."
        public ClassB() :
            base() {
        }
    }
    ```

2. Multi-line casting (often automatically done by tools like Resharper) e.g.
    ```C#
    var aValue =
        someFlag ?
        /// Violation here as "Operator (ALongClassName<WithTypeArgument, AndAnotherTypeArgument>) must not appear at the end of a line."
        (ALongClassName<WithTypeArgument, AndAnotherTypeArgument>)
        new ConcreteImplementationA() :
        new ConcreteImplementationB();
    ```

Now I understand that both of these can be easily rewritten to look more readable and arguably should have been when they were written. However, we break the build on stylecop violations and have a large existing codebase we don't want to fix up in one big bang as this will be time consuming.

These were not violations in the original stylecop so why are they violations now? (a bug in the original stylecop?)

I think it is an extension of the part of the rule that says "unary operators must be preceded by a single space, but must never be followed by any space" adding "_or a new line_".

Further it seems that the colon operator should be packaged in the rule "The following types of operator symbols must be surrounded by a single space on either side: colons...". So if a new-line is being considered a 'space' to cause the violation of the cast operator then the constructor spacing should be allowed.

**Comment by sharwell** [comment]
Hi @lowkay! Welcome to the project and thanks for taking the time to write this up with examples. For now, I'm going to mark this as **needs discussion** since it's not clear what approach we'll go with. Here is some additional information:

> These were not violations in the original stylecop so why are they violations now?

There are many reasons why slight changes were made to various rules, including:

* Consistency with other rules
* Bugs in the original implementation
* Oversights in the original implementation (e.g. the `:` you mentioned in the first part)
* Failure to account for newer language features

Our goal is to make sure that code fixes are provided in cases where we differ from StyleCop classic, avoid conflicting definitions where possible, and document the changes (#359).

> However, we break the build on stylecop violations and have a large existing codebase we don't want to fix up in one big bang as this will be time consuming.

SA1003 includes a highly accurate code fix that can be applied to an entire solution at once. I've never seen it fail to fix all violations in one pass. :smile:

**Comment by sharwell** [comment]
My vote in both of the above cases is to not change our current behavior, and to verify that our unit tests include tests for code fixes in each of these cases.

**Comment by lowkay** [comment]
Hi @sharwell that's fair enough, I agree that style-wise it's neater, if the constructor case is a violation would you consider the following as a violation? (because currently it isn't)
    ```C#
    public classB :
              classA {
    }
    ```

**Comment by sharwell** [comment]
@lowkay I modified my votes. I had forgotten that we have already approved a new rule specifically for the `:` placement.

**Comment by sharwell** [comment]
:memo: SA1128 has now been implemented.
